man/coqdep.1
.TH COQ 1

.SH NAME
coqdep \- Compute inter-module dependencies for Coq programs

.SH SYNOPSIS
.B coqdep
[
.BI \-I \ directory
]
[
.BI \-coqlib \ directory
]
[
.BI \-i
]
[
.BI \-slash
]
.I filename ...
.I directory ...

.SH DESCRIPTION

.B coqdep
computes inter-module dependencies for Coq programs,
and prints the dependencies on the standard output in a format
readable by make.
When a directory is given as argument, it is recursively looked at.

Dependencies of Coq modules are computed by looking at
.IR Require \&
commands (Require, Require Export, Require Import, possibly restricted by a From clause),
.IR Declare \&
.IR ML \&
.IR Module \&
commands,
.IR Add \&
.IR LoadPath \&
commands and
.IR Load \&
commands. Dependencies relative to modules from the Coq library are not
printed except if
.BR \-boot \&
is given.

.SH OPTIONS

.TP
.BI \-f \ file
Read filenames and options -I, -R and -Q from a _CoqProject FILE.
.TP
.BI \-I/\-Q/\-R \ options
Have the same effects on load path and modules names as for other
Coq commands (coqtop, coqc).
.TP
.BI \-coqlib \ directory
Indicates where is the Coq library. The default value has been
determined at installation time, and therefore this option should not
be used under normal circumstances.
.TP
.BI \-exclude-dir \ dir
Skips subdirectory
.IR dir \ during
.BR -R/-Q \ search.
.TP
.B \-sort
Output the given file name ordered by dependencies.
.TP
.B \-vos
Output dependencies for .vos files (this is not the default as it breaks dune's Coq mode)
.TP
.B \-boot
For coq developers, prints dependencies over coq library files
(omitted by default).
.TP
.B \-noinit
Currently no effect.
.TP
.B \-vos
Includes dependencies about .vos files.
.TP
.B \-dyndep (opt|byte|both|no|var)
Set how dependencies over ML modules are printed.

.SH EXIT STATUS
.IP "1"
A file given on the command line cannot be found, or some file
cannot be opened, or there is a syntax error in one of the commands recognized by coqdep.
.IP "0"
In all other cases. In particular, when a dependency cannot be
found or an invalid option is encountered,
.B coqdep
prints a warning and exits with status
.B 0\fR.

.TP
.B \-compute_missing
If coqdep cannot find any physical .v or .vo file matching a Require Import
of a logical path, say L, coqdep will compute the expected physical path
of the .vo file corresponding to L based on the -R/-Q directives and $(COQPATH).
That .vo file will be added as a dependency. This is useful, for example,
when the corresponding .v file should be lazily autogenerated, only when needed
by another .v file, say N.v, e.g. in a fast partial build just to interactively
edit N.v.
In such cases, the autogenerated .v files should not be added to _CoqProject
because those files are required to be eagerly present before coqdep gets invoked.
Because the contents of the autogenerated .v file are not known, it is the user's
responsibility to add rules to Makefile.local to compile
those autogenerated .v files to .vo files.
Require Imports of missing files should be fully qualified and satisfy the following
condition: Suppose a file /d1/d2/d3..dn/f.v is missing and
directories dk/../dn also don't exist.
The user needs to make sure their -R/-Q directives are such that if any logical
path L corresponds to /d1/d2/..dk-1, no other physical path (say O) corresponds to L.
Otherwise coqdep may incorrectly add a dependency on the file O/dk/../dn/f.v


.SH SEE ALSO

.BR ocamlc (1),
.BR coqc (1),
.BR make (1).
.br

.SH NOTES

Lexers (for Coq and OCaml) correctly handle nested comments
and strings.

The treatment of symbolic links is primitive.

If two files have the same name, in two different directories,
a warning is printed on standard error.

There is no way to limit the scope of the recursive search for
directories.

.SH EXAMPLES

.LP
Consider the files (in the same directory):

	a.mllib X.v Y.v and Z.v

where
.TP
.BI \+
a.mllib contains the module names `B' and `C';
.TP
.BI \+
Y.v contains the command `Require Foo.X' ;
.TP
.BI \+
Z.v contains the commands `From Foo Require X' and `Declare ML Module "a"'.
.LP
To get the dependencies of the Coq files:
.IP
.B
example% coqdep \-I . -Q . Foo *.v
.RS
.sp .5
.nf
.B X.vo X.glob X.v.beautified X.required_vo: X.v
.B X.vio: X.v
.B Y.vo Y.glob Y.v.beautified Y.required_vo: Y.v X.vo
.B Y.vio: Y.v X.vio
.B Z.vo Z.glob Z.v.beautified Z.required_vo: Z.v X.vo ./a.cma ./a.cmxs
.B Z.vio: Z.v X.vio ./a.cma ./a.cmxs
.fi
.RE
.br

.SH BUGS

Please report any bug to
.B https://github.com/coq/coq/issues


tools/coqdep.ml
(************************************************************************)
(*         *   The Coq Proof Assistant / The Coq Development Team       *)
(*  v      *         Copyright INRIA, CNRS and contributors             *)
(* <O___,, * (see version control and CREDITS file for authors & dates) *)
(*   \VV/  **************************************************************)
(*    //   *    This file is distributed under the terms of the         *)
(*         *     GNU Lesser General Public License Version 2.1          *)
(*         *     (see LICENSE file for the text of the license)         *)
(************************************************************************)

open Format
open Minisys
open Coqdep_common

(** The basic parts of coqdep (i.e. the parts used by [coqdep -boot])
    are now in [Coqdep_common]. The code that remains here concerns
    the other options. Calling this complete coqdep with the [-boot]
    option should be equivalent to calling [coqdep_boot].

    As of today, this module depends on the following Coq modules:

    - Envars
    - CoqProject_file

    All of it for `coqlib` handling. Ideally we would like to clean
    coqlib handling up so this can be bootstrapped earlier.
*)

let option_sort = ref false

let usage () =
  eprintf " usage: coqdep [options] <filename>+\n";
  eprintf " options:\n";
  eprintf "  -boot : For coq developers, prints dependencies over coq library files (omitted by default).\n";
  eprintf "  -sort : output the given file name ordered by dependencies\n";
  eprintf "  -noglob | -no-glob : \n";
  eprintf "  -f file : read -I, -Q, -R and filenames from _CoqProject-formatted FILE.";
  eprintf "  -I dir : add (non recursively) dir to ocaml path\n";
  eprintf "  -R dir logname : add and import dir recursively to coq load path under logical name logname\n";
  eprintf "  -Q dir logname : add (recursively) and open (non recursively) dir to coq load path under logical name logname\n";
  eprintf "  -vos : also output dependencies about .vos files\n";
  eprintf "  -exclude-dir dir : skip subdirectories named 'dir' during -R/-Q search\n";
  eprintf "  -coqlib dir : set the coq standard library directory\n";
  eprintf "  -dyndep (opt|byte|both|no|var) : set how dependencies over ML modules are printed\n";
  exit 1

let split_period = Str.split (Str.regexp (Str.quote "."))

let add_q_include path l = add_rec_dir_no_import add_known path (split_period l)
let add_r_include path l = add_rec_dir_import add_known path (split_period l)

let treat_coqproject f =
  let open CoqProject_file in
  let iter_sourced f = List.iter (fun {thing} -> f thing) in
  let warning_fn x = coqdep_warning "%s" x in
  let project = read_project_file ~warning_fn f in
  iter_sourced (fun { path } -> add_caml_dir path) project.ml_includes;
  iter_sourced (fun ({ path }, l) -> add_q_include path l) project.q_includes;
  iter_sourced (fun ({ path }, l) -> add_r_include path l) project.r_includes;
  iter_sourced (fun f -> treat_file None f) (all_files project)

let rec parse = function
  | "-boot" :: ll -> option_boot := true; parse ll
  | "-sort" :: ll -> option_sort := true; parse ll
  | "-compute_missing" :: ll -> option_compute_missing := true; parse ll
  | "-vos" :: ll -> write_vos := true; parse ll
  | ("-noglob" | "-no-glob") :: ll -> option_noglob := true; parse ll
  | "-f" :: f :: ll -> treat_coqproject f; parse ll
  | "-I" :: r :: ll -> add_caml_dir r; parse ll
  | "-I" :: [] -> usage ()
  | "-R" :: r :: ln :: ll -> add_r_include r ln; parse ll
  | "-Q" :: r :: ln :: ll -> add_q_include r ln; parse ll
  | "-R" :: ([] | [_]) -> usage ()
  | "-exclude-dir" :: r :: ll -> System.exclude_directory r; parse ll
  | "-exclude-dir" :: [] -> usage ()
  | "-coqlib" :: r :: ll -> Envars.set_user_coqlib r; parse ll
  | "-coqlib" :: [] -> usage ()
  | "-dyndep" :: "no" :: ll -> option_dynlink := No; parse ll
  | "-dyndep" :: "opt" :: ll -> option_dynlink := Opt; parse ll
  | "-dyndep" :: "byte" :: ll -> option_dynlink := Byte; parse ll
  | "-dyndep" :: "both" :: ll -> option_dynlink := Both; parse ll
  | "-dyndep" :: "var" :: ll -> option_dynlink := Variable; parse ll
  | ("-h"|"--help"|"-help") :: _ -> usage ()
  | f :: ll -> treat_file None f; parse ll
  | [] -> ()

(* Exception to be raised by Envars *)
exception CoqlibError of string

let coqdep () =
  if Array.length Sys.argv < 2 then usage ();
  if not Coq_config.has_natdynlink then option_dynlink := No;
  parse (List.tl (Array.to_list Sys.argv));
  (* Add current dir with empty logical path if not set by options above. *)
  (try ignore (Coqdep_common.find_dir_logpath (Sys.getcwd()))
   with Not_found -> add_norec_dir_import add_known "." []);
  (* We don't setup any loadpath if the -boot is passed *)
  if not !option_boot then begin
    Envars.set_coqlib ~fail:(fun msg -> raise (CoqlibError msg));
    let coqlib = Envars.coqlib () in
    add_rec_dir_import add_coqlib_known (coqlib//"theories") ["Coq"];
    add_rec_dir_import add_coqlib_known (coqlib//"plugins") ["Coq"];
    let user = coqlib//"user-contrib" in
    if Sys.file_exists user then add_rec_dir_no_import add_coqlib_known user [];
    List.iter (fun s -> add_rec_dir_no_import add_coqlib_known s [])
      (Envars.xdg_dirs ~warn:(fun x -> coqdep_warning "%s" x));
    List.iter (fun s -> add_rec_dir_no_import add_coqlib_known s []) Envars.coqpath;
  end;
  if !option_sort then begin sort (); exit 0 end;
  coq_dependencies ();
  ()

let _ =
  try
    coqdep ()
  with CoqlibError msg ->
    eprintf "*** Error: %s@\n%!" msg;
    exit 1
